src/natcap/invest/urban_nature_access.py

```diff
@@ -2528,6 +2527,40 @@ def _create_mask(*raster_arrays):
         _create_mask, target_mask_path, gdal.GDT_Byte, nodata_target=255)
 
 
+def _warp_lulc(source_lulc_path, target_lulc_path, target_pixel_size,
+               target_bounding_box):
+    """Warp a LULC raster and set a nodata if needed.
+
+    Args:
+        source_lulc_path (str): The path to a source LULC raster.
+        target_lulc_path (str): The path to the new LULC raster.
+        target_pixel_size (tuple): A 2-tuple of the target pixel size.
+        target_bounding_box (tuple): A 4-tuple of the target bounding box.
+
+    Returns:
+        ``None``.
+    """
+    source_raster_info = pygeoprocessing.get_raster_info(source_lulc_path)
+    target_nodata = source_raster_info['nodata'][0]
+    if target_nodata is None:
+        # Guarantee that our nodata cannot be represented by the datatype -
+        # select a nodata value that's out of range.
+        target_nodata = pygeoprocessing.choose_nodata(
+            source_raster_info['numpy_type']) + 1
```

Comment:
I'm curious about the need to add 1 here. I mean, I understand from the comment why you've done this, but I wonder if you think `choose_nodata` should be updated to return the max + 1 instead of just the max? Would that make sense for other use cases, or could it cause problems?

Reply:
Great question!  Since we can assume that every pixel in the dataset is valid (because `target_nodata is None`), the safest possible nodata value is one that isn't even in the range of values possible with the given datatype.  For a `float32` raster, it's pretty unlikely that we would have a dataset that would have every discrete value that is allowed by the float32 spec, but it could still happen.  Since the nodata value is stored as a piece of metadata separate from the pixel values themselves, we can avoid this possibility completely by using a nodata value that is completely outside of the range of possibilities of the datatype of the pixel values themselves.

If we were to change the datatype of `choose_nodata`, you are right that that would unfortunately break things.  In `raster_map`, the nodata values of the incoming raster are transformed to the nodata values of the target raster, where the target nodata has been selected by `choose_nodata`.  If `choose_nodata` were to return a value of the (dtype's max value)+1, then we'd get an overflow in the pixel values, or we bump up the output raster's datatype to accommodate the extra needed precision, neither of which we want!

So this solution as implemented gets around the situation by only handling this one case, where the lulc doesn't have a nodata value defined, and we just set one (to avoid breaking other things later on in the model) that can never happen in this datatype.